system/jlib/jdebug.cpp

```diff
@@ -818,9 +739,206 @@ struct KERNEL_USER_TIMES {
 //    );
 
 
-typedef LONG (WINAPI *PROCNTQSI)(UINT,PVOID,ULONG,PULONG);
-typedef LONG (WINAPI *PROCNTQIP)(HANDLE,UINT,PVOID,ULONG,PULONG);
-typedef LONG (WINAPI *PROCNTGST)(LARGE_INTEGER*, LARGE_INTEGER*, LARGE_INTEGER*);
+typedef LONG(WINAPI *PROCNTQIP)(HANDLE, UINT, PVOID, ULONG, PULONG);
+
+static struct CNtKernelInformation
+{
+    CNtKernelInformation()
+    {
+        NtQueryInformationProcess = (PROCNTQIP)GetProcAddress(
+            GetModuleHandle("ntdll"),
+            "NtQueryInformationProcess"
+        );
+        GetSystemInfo(&SysBaseInfo);
+}
+
+    PROCNTQIP NtQueryInformationProcess;
+
+    SYSTEM_INFO       SysBaseInfo;
+
+} NtKernelFunctions;
+#endif
+
+
+//===========================================================================
+
+#ifdef _WIN32
+static __uint64 ticksToNs = I64C(100);  // FILETIME is in 100ns increments
+#else
+static __uint64 ticksToNs = I64C(1000000000) / sysconf(_SC_CLK_TCK);
+#endif
+
+
+#ifdef _WIN32
+inline unsigned __int64 getFILETIME(const FILETIME & value)
+{
+    return ((__uint64)value.dwHighDateTime << (sizeof(value.dwLowDateTime) * 8) | value.dwLowDateTime);
+}
+#endif
+
+CpuInfo::CpuInfo(bool processTime, bool systemTime) : CpuInfo()
+{
+    if (processTime)
+        getProcessTimes();
+    else if (systemTime)
+        getSystemTimes();
+}
+
+void CpuInfo::clear()
+{
+    user = 0;
+    system = 0;
+    idle = 0;
+    iowait = 0;
+}
+
+CpuInfo CpuInfo::operator - (const CpuInfo & rhs) const
+{
+    CpuInfo result;
+    result.user = user - rhs.user;
+    result.system = system - rhs.system;
+    result.idle = idle - rhs.idle;
+    result.iowait = iowait - rhs.iowait;
+    result.ctx = ctx - rhs.ctx;
+    return result;
+}
+
+bool CpuInfo::getProcessTimes()
+{
+#ifdef _WIN32
+    FILETIME kernelTime, userTime;
+    GetProcessTimes(GetCurrentProcess(), nullptr, nullptr, &kernelTime, &userTime);
+
+    user = getFILETIME(userTime);
+    system = getFILETIME(kernelTime);
+    return true;
+#else
+    VStringBuffer fname("/proc/%u/stat", getpid());
+    //NOTE: This file needs to be reopened each time - seeking to the start and rereading does not refresh it
+    clear();
+    FILE* cpufp = fopen(fname.str(), "r");
+    if (!cpufp) {
+        return false;
+    }
+    char ln[2560];
+    if (fgets(ln, sizeof(ln), cpufp))
+    {
+        long int majorFaults = 0;
+        long unsigned userTime = 0;
+        long unsigned systemTime = 0;
+        int matched = sscanf(ln, "%*d %*s %*c %*d %*d %*d %*d %*d %*u %*u %*u %lu %*u %lu %lu", &majorFaults, &userTime, &systemTime);
+        if (matched >= 3)
+        {
+            user = userTime;
```

Comment:
Would it be better to add user/system time for child processes here?  On reflection I think yes.